zydeco-lang/src/statics/tyck/lub.rs

```diff
@@ -112,41 +125,49 @@ impl Lub for Type {
                 let body = lhs.body.inner_clone().lub(rhs.body.inner_clone(), ctx, span)?;
                 Ok(TypeAbs { params: lhs.params, body: rc!(lhs.body.span().make(body)) }.into())
             }
-            (SynType::TypeApp(TypeApp { tvar: NeutralVar::Var(tvar), args: _ }), _)
```

Comment:
Perhaps the only important thing in this PR: fix how `TypeApp` is dealt with in `Type::lub`.

Reply:
Does `Type::lub` here stand for `least upper bound`?

Reply:
A quick recap: the least upper bound was introduced to provide the ability to reason about holes in bi-directional type checking. It's an operation that takes two normalized types named `A` and `B` and inductively checks whether a. they have the same structure and b. they only differ on leaves by compatible holes. The operation also returns a more precise type.

The base cases are
```
_ \/ A = A
A \/ _ = A
```

And if neither `A` nor `B`  have `_` (hole) at the top level, check whether the scrutinees are the same and inductively get their children's lub.